packages/flutter/lib/src/widgets/editable_text.dart

```diff
@@ -3873,3 +3880,243 @@ class _CopySelectionAction extends ContextAction<CopySelectionTextIntent> {
   @override
   bool get isActionEnabled => state._value.selection.isValid && !state._value.selection.isCollapsed;
 }
+
+/// A void function that takes a [TextEditingValue].
+@visibleForTesting
+typedef TextEditingValueCallback = void Function(TextEditingValue value);
+
+/// Provides undo/redo capabilities for text editing.
+///
+/// Listens to [controller] as a [ValueNotifier] and saves relevant values for
+/// undoing/redoing. The cadence at which values are saved is a best
+/// approximation of the native behaviors of a hardware keyboard on Flutter's
+/// desktop platforms, as there are subtle differences between each of these
+/// platforms.
+///
+/// Listens for keyboard undo/redo shortcuts depending on the platform. Calls
+/// [onChanged] when a shortcut is received that would change the state of these
+/// [controller].
+@visibleForTesting
+class UndoRedo extends StatefulWidget {
+  /// Creates an instance of [UndoRedo].
+  const UndoRedo({
+    Key? key,
+    required this.child,
+    required this.controller,
+    required this.onChanged,
+  }) : super(key: key);
+
+  /// The child widget of [UndoRedo].
+  final Widget child;
+
+  /// The [ValueNotifier] to save the state of over time.
+  final TextEditingController controller;
+
+  /// Called when an undo or redo is received, even if the state would be the
+  /// same.
+  final TextEditingValueCallback onChanged;
+
+  @override
+  UndoRedoState createState() => UndoRedoState();
```

Comment:
```suggestion
  State<UndoRedo> createState() => _UndoRedoState();
```

Reply:
Good catch, that's why I should autogenerate these things.